### PR TITLE
Deployment contracts without review

### LINKED ISSUE #3379: Deployment contracts without review
- Add ACS12 for user contract.
- Contracts that implement ACS12 do not need to be reviewed at contract deployment time.

### PR DESCRIPTION
**Pull request by zhxymh**
Closes #3379

### CODE REVIEW

File: src/AElf.Kernel.CodeCheck/Application/ICodeCheckProposalService.cs [hunk 1]
```diff
@@ -0,0 +1,10 @@
+using System.Collections.Generic;
+
+namespace AElf.Kernel.CodeCheck.Application;
+
+public interface ICodeCheckProposalService
+{
+    void AddToReleasedProposal(Hash proposalId, Hash proposalInputHash, long height);
```
Comment (gldeng) [inline src/AElf.Kernel.CodeCheck/Application/ICodeCheckProposalService.cs:7]: The naming doesn't follow English grammar. Maybe change to `AddReleasableProposal`.

File: src/AElf.Kernel.CodeCheck/Application/ICodeCheckProposalService.cs [hunk 2]
```diff
@@ -0,0 +1,10 @@
...
+    Task<List<CodeCheckProposal>> GetToReleasedProposalListAsync(Address from, Hash blockHash, long blockHeight);
```
Comment (gldeng) [inline src/AElf.Kernel.CodeCheck/Application/ICodeCheckProposalService.cs:8]: Maybe change to `GetReleasableProposalListAsync`.

File: src/AElf.Kernel.CodeCheck/Application/ICodeCheckProposalService.cs (hunk 1 above)
Comment (zhxymh) [reply src/AElf.Kernel.CodeCheck/Application/ICodeCheckProposalService.cs]: Got it.

File: src/AElf.Kernel.CodeCheck/Application/ICodeCheckProposalService.cs (hunk 2 above)
Comment (zhxymh) [reply src/AElf.Kernel.CodeCheck/Application/ICodeCheckProposalService.cs]: Got it.

File: contract/AElf.Contracts.MultiToken/TokenContract_Fees.cs [hunk 3]
```diff
@@ -33,118 +30,222 @@ public override ChargeTransactionFeesOutput ChargeTransactionFees(ChargeTransact
...
+        // Primary token not created yet.
+        return State.ChainPrimaryTokenSymbol.Value == null;
+    }
+
+    private UserContractMethodFees GetActualFee(Address contractAddress, string methodName)
+    {
+        if (State.ConfigurationContract.Value == null)
+            State.ConfigurationContract.Value =
+                Context.GetContractAddressByName(SmartContractConstants.ConfigurationContractSystemName);
+        var spec = State.ConfigurationContract.GetConfiguration.Call(new StringValue
```
Comment (gldeng) [inline contract/AElf.Contracts.MultiToken/TokenContract_Fees.cs:110]: Can add some  comments to explain what's going on here?

File: contract/AElf.Contracts.MultiToken/TokenContract_Fees.cs [hunk 4]
```diff
@@ -33,118 +30,222 @@ public override ChargeTransactionFeesOutput ChargeTransactionFees(ChargeTransact
...
+            Value = $"{TokenContractConstants.UserContractMethodFeeKey}_{contractAddress}_{methodName}"
+        });
+        var fee = new UserContractMethodFees();
+        if (!spec.Value.IsNullOrEmpty())
+        {
+            fee.MergeFrom(spec.Value);
+            return fee;
+        }
+
+        var value = State.ConfigurationContract.GetConfiguration.Call(new StringValue
```
Comment (gldeng) [inline contract/AElf.Contracts.MultiToken/TokenContract_Fees.cs:121]: And here?

File: contract/AElf.Contracts.MultiToken/TokenContract_Fees.cs [hunk 5]
```diff
@@ -183,23 +284,43 @@ private void SetOrRefreshMethodFeeFreeAllowances(Address address)
...
+                dict[methodFee.Symbol] = dict[methodFee.Symbol].Add(methodFee.BasicFee);
+            }
+            else
+            {
+                dict[methodFee.Symbol] = methodFee.BasicFee;
+            }
+        }
+
+        return dict;
+    }
```
Comment (gldeng) [inline contract/AElf.Contracts.MultiToken/TokenContract_Fees.cs:303]: Consider refactor to:
```C#
    private Dictionary<string, long> GetBaseFeeDictionary(MethodFees methodFees)
    {
        return methodFees.Fees
            .GroupBy(f => f.Symbol, f => f.BasicFee)
            .ToDictionary(g => g.Key, g => g.Sum());
    }
```

File: protobuf/acs12.proto [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+
+import public "aelf/options.proto";
+import public "google/protobuf/empty.proto";
+import public "google/protobuf/wrappers.proto";
+import "aelf/core.proto";
+
+option (aelf.identity) = "acs12";
+option csharp_namespace = "AElf.Standards.ACS12";
+
+service UserContractMethodFeeProviderContract {
```
Comment (gldeng) [inline protobuf/acs12.proto:18]: `acs12` is not only used to provide MethodFee but also used to denote the UserContract (which requires CodeCheck). So is it better to call it `UserContract`?

File: contract/AElf.Contracts.MultiToken/TokenContract_Fees.cs [hunk 7]
```diff
@@ -33,118 +30,222 @@ public override ChargeTransactionFeesOutput ChargeTransactionFees(ChargeTransact
...
+        var fee = new UserContractMethodFees();
+        if (!spec.Value.IsNullOrEmpty())
+        {
+            fee.MergeFrom(spec.Value);
+            return fee;
+        }
+
+        var value = State.ConfigurationContract.GetConfiguration.Call(new StringValue
+        {
+            Value = TokenContractConstants.UserContractMethodFeeKey
```
Comment (gldeng) [inline contract/AElf.Contracts.MultiToken/TokenContract_Fees.cs:123]: 没明白这个配置是谁去设置。。。

File: contract/AElf.Contracts.MultiToken/TokenContract_Fees.cs [hunk 8]
```diff
@@ -33,118 +30,222 @@ public override ChargeTransactionFeesOutput ChargeTransactionFees(ChargeTransact
...
+    }
+
+    private UserContractMethodFees GetActualFee(Address contractAddress, string methodName)
+    {
+        if (State.ConfigurationContract.Value == null)
+            State.ConfigurationContract.Value =
+                Context.GetContractAddressByName(SmartContractConstants.ConfigurationContractSystemName);
+        var spec = State.ConfigurationContract.GetConfiguration.Call(new StringValue
+        {
+            Value = $"{TokenContractConstants.UserContractMethodFeeKey}_{contractAddress}_{methodName}"
```
Comment (gldeng) [inline contract/AElf.Contracts.MultiToken/TokenContract_Fees.cs:112]: 没明白这个配置是谁去设置。。。

File: src/AElf.Kernel.CodeCheck/Application/CodeCheckProposalService.cs [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+
+    public void AddReleasableProposal(Hash proposalId, Hash proposalInputHash, long height)
+    {
+        _codeCheckProposalProvider.AddProposal(proposalId, proposalInputHash, height);
+    }
+
+    public async Task<List<CodeCheckProposal>> GetReleasableProposalListAsync(Address from, Hash blockHash,
+        long blockHeight)
+    {
+        var proposalList = _codeCheckProposalProvider.GetAllProposals();
```
[the remark this answers is not in the document]
Comment (zhxymh) [reply src/AElf.Kernel.CodeCheck/Application/CodeCheckProposalService.cs]: Done.

File: src/AElf.Kernel.CodeCheck/Application/CodeCheckProposalService.cs [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+                ContractAddress = await GetParliamentContractAddressAsync(new ChainContext
+                {
+                    BlockHash = blockHash,
+                    BlockHeight = blockHeight
+                })
+            }).GetAvailableProposals
+            .CallAsync(new ProposalIdList { ProposalIds = { proposalIdList } });
+
+        if (result == null)
+            return;
```
[the remark this answers is not in the document]
Comment (zhxymh) [reply src/AElf.Kernel.CodeCheck/Application/CodeCheckProposalService.cs]: Removed.

File: src/AElf.Kernel.CodeCheck/Application/CodeCheckProposalReleaseTransactionGenerator.cs [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+            {
+                BlockHash = preBlockHash,
+                BlockHeight = preBlockHeight
+            }, proposal.ProposalId);
+            
+            Logger.LogTrace("Code check proposal release transaction generated: {proposalId}.",proposal.ProposalId.ToHex());
+        }
+        
+        return generatedTransactions;
+    }
```
[the remark this answers is not in the document]
Comment (zhxymh) [reply src/AElf.Kernel.CodeCheck/Application/CodeCheckProposalReleaseTransactionGenerator.cs]: Done.

File: protobuf/acs12.proto (hunk 6 above)
Comment (zhxymh) [reply protobuf/acs12.proto]: Done.

**Review by gldeng — APPROVED** [review]
LGTM

**Review by jason-hoopox — APPROVED** [review]

**Pull request merged**